transaction-history-manager/src/main/java/com/qpay/transactionhistorymanager/service/impl/PdfReportGeneratorITextPdf.java

```diff
@@ -0,0 +1,217 @@
+
+    public void generateCustomerPdfReport(final PdfReportInfo pdfReportInfo) throws IOException {
+        final CustomerReportInfo customerReportInfo = userClient.getCustomerReportInfo(pdfReportInfo.userId());
+
+        final String path = generatePath(pdfReportInfo.userId());
+        BigDecimal totalReceived = BigDecimal.valueOf(0);
+        BigDecimal totalSent = BigDecimal.valueOf(0);
+        final Table transactionsTable = createTransactionsTable();
+        final List<TransactionEntity> transactionEntitiesInPeriod = transactionRepository
+                .findAllByIdFromAndCreatedAtBetween(
+                        pdfReportInfo.userId(),
+                        pdfReportInfo.periodStart(),
+                        pdfReportInfo.periodEnd()
+                );
+        for (int number = 0; number < transactionEntitiesInPeriod.size(); number++) {
+            final TransactionEntity transactionEntity = transactionEntitiesInPeriod.get(number);
+            final BigDecimal amount = transactionEntity.getAmount();
+            final TransactionType transactionType = transactionEntity.getTransactionType();
+
+            transactionsTable.addCell(String.valueOf(number + 1));
+            transactionsTable.addCell(transactionEntity.getCreatedAt()
+                    .format(DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm")
+            ));
+            transactionsTable.addCell(transactionType.toString());
+            transactionsTable.addCell(amount.toString());
+            transactionsTable.addCell(transactionEntity.getNameTo());
+
+            if (transactionType == TransactionType.PAYMENT) {
+                totalSent = totalSent.add(amount);
+            } else if (transactionType == TransactionType.TOP_UP) {
+                totalReceived = totalReceived.add(amount);
+            }
+        }
+
+        final List<Paragraph> paragraphList = List.of(
+                createLogoParagraph(),
+                createTitleParagraph(),
+                createSubTitleParagraph(pdfReportInfo),
+                createClientNameParagraph(customerReportInfo.name()),
+                createTotalPaymentsReceivedParagraph(totalReceived),
+                createTotalPaymentsSentParagraph(totalSent),
+                new Paragraph().add(transactionsTable)
+        );
+        FileManager.savePdf(path, paragraphList);
+    }
+
+    public void generateMerchantPdfReport(final PdfReportInfo pdfReportInfo) throws IOException {
+        final MerchantReportInfo merchantReportInfo = userClient.getMerchantReportInfo(pdfReportInfo.userId());
+
+        final String path = generatePath(pdfReportInfo.userId());
+        BigDecimal totalReceived = BigDecimal.valueOf(0);
+        BigDecimal totalSent = BigDecimal.valueOf(0);
+        final Table transactionsTable = createTransactionsTable();
+        final List<TransactionEntity> transactionEntitiesInPeriod = transactionRepository
+                .findAllByIdToAndCreatedAtBetween(
```

Comment:
we don't have customer to customer, merchant to merchant and merchant to customer payment. idFrom is for customer and idTo is for merchant. Otherwise we have to add userTypeFrom and userTypeTo in transactionEntity, because ids can be the same.

Reply:
@adsukharev